core/src/latest_unprocessed_votes.rs

```diff
@@ -0,0 +1,700 @@
+    solana_perf::packet::{Packet, PacketBatch},
+    solana_runtime::bank::Bank,
+    solana_sdk::{clock::Slot, program_utils::limited_deserialize, pubkey::Pubkey},
+    solana_vote_program::vote_instruction::VoteInstruction,
+    std::{
+        cell::RefCell,
+        collections::HashMap,
+        ops::DerefMut,
+        rc::Rc,
+        sync::{Arc, RwLock},
+    },
+};
+
+#[derive(PartialEq, Eq, Debug, Copy, Clone)]
+pub enum VoteSource {
+    Gossip,
+    Tpu,
+}
+
+/// Holds deserialized vote messages as well as their source, foward status and slot
+#[derive(Debug, Clone)]
+pub struct DeserializedVotePacket {
+    vote_source: VoteSource,
+    pubkey: Pubkey,
+    vote: Option<Rc<ImmutableDeserializedPacket>>,
+    slot: Slot,
+    forwarded: bool,
+}
+
+impl DeserializedVotePacket {
+    pub fn new(packet: Packet, vote_source: VoteSource) -> Result<Self, DeserializedPacketError> {
+        if !packet.meta.is_simple_vote_tx() {
+            return Err(DeserializedPacketError::VoteTransactionError);
+        }
+
+        let vote = Rc::new(ImmutableDeserializedPacket::new(packet, None)?);
+        Self::new_from_immutable(vote, vote_source)
+    }
+
+    pub fn new_from_immutable(
+        vote: Rc<ImmutableDeserializedPacket>,
+        vote_source: VoteSource,
+    ) -> Result<Self, DeserializedPacketError> {
+        let message = vote.transaction().get_message();
+        let (_, instruction) = message
+            .program_instructions_iter()
+            .next()
+            .ok_or(DeserializedPacketError::VoteTransactionError)?;
+
+        match limited_deserialize::<VoteInstruction>(&instruction.data) {
+            Ok(VoteInstruction::UpdateVoteState(vote_state_update))
+            | Ok(VoteInstruction::UpdateVoteStateSwitch(vote_state_update, _)) => {
+                let &pubkey = message
+                    .message
+                    .static_account_keys()
+                    .get(0)
+                    .ok_or(DeserializedPacketError::VoteTransactionError)?;
+                let slot = vote_state_update.last_voted_slot().unwrap_or(0);
+
+                Ok(Self {
+                    vote: Some(vote),
+                    slot,
+                    pubkey,
+                    vote_source,
+                    forwarded: false,
+                })
+            }
+            _ => Err(DeserializedPacketError::VoteTransactionError),
+        }
+    }
+
+    pub fn get_vote_packet(&self) -> Rc<ImmutableDeserializedPacket> {
+        self.vote.as_ref().unwrap().clone()
+    }
+
+    pub fn pubkey(&self) -> Pubkey {
+        self.pubkey
+    }
+
+    pub fn slot(&self) -> Slot {
+        self.slot
+    }
+
+    pub fn is_forwarded(&self) -> bool {
+        // By definition all gossip votes have been forwarded
+        self.forwarded || matches!(self.vote_source, VoteSource::Gossip)
+    }
+
+    pub fn is_processed(&self) -> bool {
+        self.vote.is_none()
+    }
+
+    pub fn clear(&mut self) {
```

Comment:
nit: rename to `take_vote` and return a `Option<Rc<ImmutableDeserializedPacket>>` via `self.vote.take()`